### DISCUSSION TITLE
Are there any getStaticProps/getServerSideProps styled-component examples?

### DISCUSSION THREAD

**Discussion by TheoMer** [vercel/next.js#22065] [Help]
Are there any styled-component examples of _app.js and _document.js using getServerSideProps instead of getInitialProips available as shown here?:

https://github.com/vercel/next.js/tree/canary/examples/with-styled-components/pages

**Comment by trompx** [comment]
I am also wondering the same. @timneutkens What is the point to use styled components with next.js if it completely prevents static optimisation for the whole app by using getInitialProps in _document.js?

**Reply by j3ll3nl** [reply]
It's not true that it prevents optimizing.
Check https://nextjs.org/docs/advanced-features/custom-document

It seems that using getInitialProps there is still best practice.

**Reply by isaachinman** [reply]
@j3ll3nl The `getInitialProps` in that example is not necessary unless you use `getInitialProps`.

**Reply by dtothefp** [reply]
I'd like to revive this conversation after the next@12 release. I'm still unclear from the documentation if using `getIntialProps` disables the ability to use some features such as static optimization and the newly released ability to use Server Components 🤷 

For example I see this in the [custom document docs](https://nextjs.org/docs/advanced-features/custom-document#caveats) about using `getInitialProps` in a custom document 

```
Document's getInitialProps function is not called during client-side transitions, nor when a page is statically optimized.
```

And then the [static optimization docs](https://nextjs.org/docs/advanced-features/automatic-static-optimization) say

```
If getServerSideProps or getInitialProps is present in a page, Next.js will switch to render the page on-demand, per-request (meaning Server-Side Rendering).

If the above is not the case, Next.js will statically optimize your page automatically by prerendering the page to static HTML.
```

From these [caveats](https://nextjs.org/docs/advanced-features/automatic-static-optimization#caveats) I'm assuming that server optimized pages are safe to use with the `document.getInitialProps` but I'm wondering if there are any other new NextJS features that are opted out of (other than SWC for styled-components which looks like has been solved in a canary)?

```
If you have a custom Document with getInitialProps be sure you check if ctx.req is defined before assuming the page is server-side rendered. ctx.req will be undefined for pages that are prerendered.
```

If I run `yarn build` and look in my `.next/server/pages/**/*.html` files it looks like the styled-components style tag is created so I'm assuming static optimization is working as intended with style-components

**Reply by joonashak** [reply]
Any update on this? Am also wondering whether I'm locked into using classes and `getInitialProps` to inject `styled-components` SSR stylesheet.

**Reply by JonScottLee** [reply]
Also wondering the same. We added this to our app to fix a fouc issue (we are using Styled Components), and following this static rendering has stopped working. Interestingly, I _do_ see the pages being rendered in the `dist` directory. But if I check the document served up by NextJS, it contains no page content -- all of that is rendered in the client.

**Reply by leerob** [reply]
https://github.com/vercel/app-playground has an example with styled-components
